packages/office-ui-fabric-react/src/components/Coachmark/Coachmark.base.tsx

```diff
@@ -310,12 +319,21 @@ export class CoachmarkBase extends BaseComponent<ICoachmarkProps, ICoachmarkStat
     return !shallowCompare(newProps, this.props) || !shallowCompare(newState, this.state);
   }
 
+  public componentWillUnmount() {
+    this._async.dispose();
+    this._disposables.forEach((dispose: () => void) => dispose());
+  }
+
   public componentDidUpdate(prevProps: ICoachmarkProps, prevState: ICoachmarkState): void {
     if (prevState.targetAlignment !== this.state.targetAlignment || prevState.targetPosition !== this.state.targetPosition) {
       this._setBeakPosition();
     }
-    if (prevProps.preventDismissOnLostFocus !== this.props.preventDismissOnLostFocus) {
+    if (prevProps.preventDismissOnLostFocus !== this.props.preventDismissOnLostFocus && !this._hasListeners) {
       this._addListeners();
+    } else {
+      if (this._hasListeners) {
+        this._removeListeners;
```

Comment:
The logic here seems to be changing a bit, is this intentional or necessary?

Also if the check is necessary you can drop the boolean in favor of checking the length of your listener array(s).

Reply:
This is intentional - I'm following the logic in CalloutContent.base.tsx where this._hasListeners is created & used to only add them if there aren't any existing & only call remove when there are existing: https://github.com/OfficeDev/office-ui-fabric-react/blob/master/packages/office-ui-fabric-react/src/components/Callout/CalloutContent.base.tsx#L105